hack/release.sh

```diff
@@ -0,0 +1,175 @@
+
+Note that this script expected
+
+Usage: hack/release.sh
+
+    -h      show this message and exit
+
+Environments:
+
+    REGISTRY                    container registry without repo name (default: quay.io/external_storage)
+    VERSION                     if set, use given version as image tag
+    DOCKER_CONFIG               optional docker config location
+    CONFIRM                     set this to skip confirmation
+    ALLOW_UNSTABLE              by default, only master branch and tags that matches v<major>.<minor>.<patch> format are allowed, set this to skip this check (debug only)
+    ALLOW_DIRTY                 by default, git repo must be clean, set this to skip this check (debug only)
+    SKIP_BUILD                  set this to skip build phase (debug only)
+
+Examples:
+
+1) Release to your own registry for testing
+
+    git tag v2.2.3
+    REGISTRY=quay.io/<yourname> ./hack/release.sh
+
+2) Release canary version
+
+    REGISTRY=quay.io/<yourname> ALLOW_UNSTABLE=true VERSION=canary ./hack/release.sh
+
+EOF
+}
+
+while getopts "h?" opt; do
+    case "$opt" in
+    h|\?)
+        usage
+        exit 0
+        ;;
+    esac
+done
+
+REGISTRY=${REGISTRY:-quay.io/external_storage}
+VERSION=${VERSION:-}
+CONFIRM=${CONFIRM:-}
+DOCKER_CONFIG=${DOCKER_CONFIG:-}
+ALLOW_UNSTABLE=${ALLOW_UNSTABLE:-}
+ALLOW_DIRTY=${ALLOW_DIRTY:-}
+SKIP_BUILD=${SKIP_BUILD:-}
+
+# remove trailing `/` if present
+REGISTRY=${REGISTRY%/}
+
+GIT_DIRTY=$(test -n "`git status --porcelain`" && echo "dirty" || echo "clean")
+
+if [ -z "$ALLOW_DIRTY" -a "$GIT_DIRTY" != "clean" ]; then
+    echo "error: repo status is not clean, skipped"
+    exit 1
+fi
+
+# our logic depends repo tags, make sure all tags are fetched
+git fetch --tags
+
+if [ -z "$VERSION" ]; then
+    echo "info: VERSION is not specified, detect automatically"
+    # get version from tag
+    VERSION=$(git describe --tags --abbrev=0 --exact-match 2>/dev/null || true)
+    if [ -z "$VERSION" ]; then
+        echo "error: failed to detect version"
+        exit 1
+    fi
+fi
+
+# By default, only v<major>.<minor>.<patch> version are allowed.
+function is_stable_version() {
+    [[ "$1" =~ ^v[0-9]+\.[0-9]+\.[0-9]+$ ]]
+}
+
+if [ -z "$ALLOW_UNSTABLE" ]; then
+    if ! is_stable_version "$VERSION"; then
+        echo "error: unstable version '$VERSION', must match regex 'v[0-9]+\.[0-9]+\.[0-9]+$' or skip this check with ALLOW_UNSTABLE"
+        exit 1
+    fi
+fi
+
+image="$REGISTRY/local-volume-provisioner:$VERSION"
+if [ -z "$CONFIRM" ]; then
+    read -r -p "info: build and push to $image? [y/N]" response
+    if [[ ! $response =~ ^([yY][eE][sS]|[yY])$ ]]; then
+        echo "Exited."
+        exit 0
+    fi
+fi
+
+# build & push container
+
+if [ -z "$SKIP_BUILD" ]; then
+    echo "info: building $image"
+    pushd provisioner &>/dev/null
+    make container REGISTRY=$REGISTRY VERSION=$VERSION
+    popd
+else
+    echo "info: building is skipped"
+fi
+
+echo "info: pushing $image"
```

Comment:
Can you add a check somewhere here that if a stable image has already been pushed, we should not override it?

Reply:
Updated. https://github.com/kubernetes-sigs/sig-storage-local-static-provisioner/pull/34/files#diff-cede583a27c227d458ca67079957c8a2R136